### DISCUSSION TITLE
A Call to Revitalize the `scikit-multilearn` Library

### DISCUSSION THREAD

**Discussion by northern-64bit** [scikit-multilearn/scikit-multilearn#304] [General]
## Critical Need for Maintenance and Enhancements
Greetings, scikit-multilearn community!

Our library, `scikit-multilearn`, is at a crucial juncture. Despite its significant role in our workflows, the latest release dates back to 2018, and as issue #257 underlines, the limited access to PyPI credentials hampers our ability to issue new releases. The accumulation of unreviewed pull requests and overall reduced activity levels pose an essential question: *What is the future of `scikit-multilearn`?*

## Proposed Solution: Establishing a New Fork
I propose we consolidate our efforts to maintain a new fork of `scikit-multilearn`. The necessity of this library is evident – a quick GitHub search for `import skmultilearn` returns over 1,900 results, showcasing a substantial and active user base.

Acknowledging the initial effort by @crypdick in the discussion #261 to start a fork, which unfortunately hasn't gained momentum, I believe it's time for us to step up and breathe new life into `scikit-multilearn`.

## Invitation for Collaboration and Discussion
This initiative is more than just a project; it's a collaborative opportunity to decide the direction of `scikit-multilearn`. Your thoughts, experiences, and potential commitment to a new fork – whether as users, contributors, or maintainers – are crucial. Also discussing questions like: who should fork it, should we create a new organization for it, or what do you think?

### Ways to Contribute, if the consensus is to make a new fork:
- Transition your PRs and solutions to the new fork, when and if created.
- Consider becoming a maintainer to help guide the library's future.
- Share your ideas and suggestions for the effective management of the new fork or whatever is decided upon.

## Join the Conversation
This moment is pivotal for `scikit-multilearn`. Together, we can ensure its continued relevance and enhancement. I apologize if this direct mention feels intrusive, but your input is vital.

## Mentions (in no particular order):
@francescodisalvo05, @erikhuck, @niedakh, @danilobiella, @ChristianSch, @Jihyunmoon0522, @makyol, @marcelopalermo, @ruslaniv, @Alex-Bujorianu, @arav-agarwal2, @Taytay, @orlevit, @zbeloki, @AlexMRuch

## Expressing Gratitude
Thank you for your time and consideration in this important matter. Let's join forces to secure the future of `scikit-multilearn`.

**Comment by ruslaniv** [comment]
I'm all for the fork. If the author is reluctant to maintain the library themselves and unwilling to accept contributions from other people, I say the proposed solution is the best one

**Comment by Alex-Bujorianu** [comment]
Hi @northern-64bit thank you for proposing this solution. I too was wondering where we go from here. With limited access to the pypi credentials I think that, indeed, the only solution is to create a fork. I’m still not sure on the name though :) In my opinion, the current name is too similar to sklearn (maybe “multilabel-sklearn”?)

**Reply by northern-64bit** [reply]
Hello @Alex-Bujorianu! I'm glad to see there's shared enthusiasm about this direction. I agree that a name change is necessary, also in order to avoid confusion with the unmaintained version and to help users easily identify the new, maintained fork. The suggestion of "multilabel-sklearn" is a great start. Other options like "sk-multilabel-learn" or "sk-multilearn" could work, but they are too similar to sklearn, potentially leading to confusion. The compatibility with scikit-learn is crucial for the package, but just due to that it doesn't need to have a nearly identical name.

**Reply by ChristianSch** [reply]
I think multilabel-sklearn is too similar. Also, the normal pattern is `sklearn-*` which is widely used with community packages. How about just `scikit-multilearn-ng`, or depending on the focus `sklearn-multilabel`.

**Reply by northern-64bit** [reply]
Since `sklearn-*` and `scikit-*` are common pattern, it would make sense to follow the convention. I also like `scikit-multilearn-ng` since it shows that this is related to `scikit-multilearn`. The naming of `sklearn-multilabel` would also be a bit to focused in my view.

So I would vote for `scikit-multilearn-ng` as the name. Please let me know if anyone is against it.

**Comment by ChristianSch** [comment]
Very good idea, you have my support. This project needs a core community of people invested in this project with a democratic process and lots of automation to help with limited maintainers. All the older core maintainers like myself probably don't use this lib anymore, so we need a next generation to take care of it. Hit me up if you have questions, all the best!

**Reply by northern-64bit** [reply]
Thank you so much for your support and encouragement! In order to get a core community with a democratic approach and robust automation to ensure the project thrives, I think that we need to start the fork under a new organisation. I'll definitely reach out if needed as this project moves forward. Thanks again, and all the best to you too!

**Comment by ghost** [comment]
I am for the fork, too. I changed some files to have the package work with Python 3.10. I am using skmultilearn in my doctoral thesis. As a workaround, I placed the modified Scikit multi-learn package in my Git Hub and loaded it to PyCharm for convenience instead of the original. For other things like GridSearchCV examples, they always return score=nan for anything (I couldn't change it yet). I can share the changes here to collaborate with the fork.

**Reply by northern-64bit** [reply]
That's great! Making the package able to run on Python 3.10 is definitely needed. So you sharing the changes would be really helpful! I'll let you know when the fork has been created, but first we'll probably have to figure out a "new" name as mentioned in https://github.com/scikit-multilearn/scikit-multilearn/discussions/304#discussioncomment-7817100 and then create an organisation, which then will create the fork.

**Comment by makyol** [comment]
I too support to fork, as it seems like the core maintainers are also ok with it. I would like the support the migration as much as I can.

**Comment by ChristianSch** [comment]
I just found some code that you folks might be interested in: https://github.com/ChristianSch/skml
I implemented ECC and PCC for my masters thesis there and if I remember correctly, is missing here. Let me know if this is something you're interested in later on.

**Reply by northern-64bit** [reply]
That's great! You're right about PCC and ECC missing in `scikit-multilearn`. They fit the package and should be added. But as you mention, this is for a later stage after the setup of the new fork and the maintenance.

**Comment by Alex-Bujorianu** [comment]
Okay @northern-64bit Can you please create an organisation and add me to it? Thanks.

**Reply by northern-64bit** [reply]
I've now created the organisation (https://github.com/scikit-multilearn-ng/) and invited you. Feel free to begin and make changes!

**Comment by crypdick** [comment]
I am happy to see this happening! Pretty soon after I started my fork ([#261](https://github.com/scikit-multilearn/scikit-multilearn/discussions/261))  I started making breaking changes and brought my dev effort into my company's private repo. Unfortunately, I cannot open source my work on stratifiers.

**Reply by northern-64bit** [reply]
No worries! It was partially thanks to you that I got the idea to revitalise this library. Feel free to contribute and if possible open PR:s with what isn't proprietary (for instance bug fixes). The current focus is to make it work and fix all of the bugs. Thanks for the support and encouragement!

**Comment by northern-64bit** [comment]
## Hello everyone!

I have now created the fork and the organisation, with the name `scikit-multilearn-ng`. Here's the link: https://github.com/scikit-multilearn-ng/scikit-multilearn-ng (Note that I didn't fork it and instead choose to import it, in order to avoid the "contribute" and "sync fork" UI. The naming and everything is still up to debate)

There's still much to be done. During the weekend I'll mainly work on getting everything sorted, for instance updating the README.md and so on.

### How to help
If anyone wants to become a maintainer, collaborator or wants to join the organisation, make sure to react to this message with an emoji (I cannot see who upvoted this) or reply. Everyone is free and encouraged to open their pull requests and contribute to the new repo, by for instance:
- Transitioning your PRs and solutions
- Setting up linting and automatic code checks
- Share your ideas and suggestions for the effective management
